File: mobile_app/app/src/main/java/com/shakealert/WebRTCHandler.kt
package com.shakealert

import android.content.Context
import android.util.Log
import org.json.JSONObject
import org.webrtc.*

class WebRTCHandler(
    private val context: Context,
    private val onOfferReady: (String) -> Unit,
    private val onIceCandidateReady: (JSONObject) -> Unit
) {
    private var peerConnectionFactory: PeerConnectionFactory? = null
    private var peerConnection: PeerConnection? = null
    private var localAudioTrack: AudioTrack? = null

    companion object {
        private const val TAG = "WebRTCHandler"
    }

    init {
        initWebRTC()
    }

    private fun initWebRTC() {
        Log.i(TAG, "Initializing WebRTC Component...")
        PeerConnectionFactory.initialize(
            PeerConnectionFactory.InitializationOptions.builder(context)
                .setEnableInternalTracer(true)
                .createInitializationOptions()
        )

        val options = PeerConnectionFactory.Options()
        val audioDeviceModule = JavaAudioDeviceModule.builder(context)
            .setUseHardwareAcousticEchoCanceler(true)
            .setUseHardwareNoiseSuppressor(true)
            .createAudioDeviceModule()

        peerConnectionFactory = PeerConnectionFactory.builder()
            .setOptions(options)
            .setAudioDeviceModule(audioDeviceModule)
            .createPeerConnectionFactory()

        audioDeviceModule.release()

        createLocalAudioTrack()
    }

    private fun createLocalAudioTrack() {
        val factory = peerConnectionFactory ?: return
        val audioSource = factory.createAudioSource(MediaConstraints())
        localAudioTrack = factory.createAudioTrack("ARDAMSa0", audioSource)
        localAudioTrack?.setEnabled(true)
        Log.i(TAG, "Local Audio Microphone Track Captured.")
    }

    fun startCall() {
        Log.i(TAG, "Initiating WebRTC Audio Pipeline (Creating Offer)...")
        if (peerConnection != null) {
            peerConnection?.close()
            peerConnection = null
        }

        val rtcConfig = PeerConnection.RTCConfiguration(
            listOf(
                PeerConnection.IceServer.builder("stun:stun.l.google.com:19302").createIceServer()
            )
        )

        peerConnection = peerConnectionFactory?.createPeerConnection(rtcConfig, object : PeerConnection.Observer {
            override fun onIceCandidate(candidate: IceCandidate?) {
                if (candidate != null) {
                    val json = JSONObject()
                    json.put("sdpMLineIndex", candidate.sdpMLineIndex)
                    json.put("sdpMid", candidate.sdpMid)
                    json.put("candidate", candidate.sdp)
                    onIceCandidateReady(json)
                }
            }
            override fun onSignalingChange(p0: PeerConnection.SignalingState?) {}
            override fun onIceConnectionChange(p0: PeerConnection.IceConnectionState?) {}
            override fun onIceConnectionReceivingChange(p0: Boolean) {}
            override fun onIceGatheringChange(p0: PeerConnection.IceGatheringState?) {}
            override fun onIceCandidatesRemoved(p0: Array<out IceCandidate>?) {}
            override fun onAddStream(p0: MediaStream?) {}
            override fun onRemoveStream(p0: MediaStream?) {}
            override fun onDataChannel(p0: DataChannel?) {}
            override fun onRenegotiationNeeded() {}
            override fun onAddTrack(p0: RtpReceiver?, p1: Array<out MediaStream>?) {}
        })

        localAudioTrack?.let {
            val mediaStreamIds = listOf("ARDAMS")
            peerConnection?.addTrack(it, mediaStreamIds)
        }

        peerConnection?.createOffer(object : SdpObserver {
            override fun onCreateSuccess(sessionDescription: SessionDescription?) {
                sessionDescription?.let { sdp ->
                    peerConnection?.setLocalDescription(object : SdpObserver {
                        override fun onCreateSuccess(p0: SessionDescription?) {}
                        override fun onSetSuccess() {
                            Log.i(TAG, "SDP Request Generated. Ready to send.")
                            onOfferReady(sdp.description)
                        }
                        override fun onCreateFailure(p0: String?) {}
                        override fun onSetFailure(p0: String?) {}
                    }, sdp)
                }
            }
            override fun onSetSuccess() {}
            override fun onCreateFailure(p0: String?) {}
            override fun onSetFailure(p0: String?) {}
        }, MediaConstraints().apply {
             mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveAudio", "true"))
             mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveVideo", "false"))
        })
    }

    fun handleRemoteAnswer(sdpDescription: String) {
        val sdp = SessionDescription(SessionDescription.Type.ANSWER, sdpDescription)
        peerConnection?.setRemoteDescription(object : SdpObserver {
            override fun onCreateSuccess(p0: SessionDescription?) {}
            override fun onSetSuccess() {
                Log.i(TAG, "Remote Rescuer Answer received! Audio Bridged Successfully.")
            }
            override fun onCreateFailure(p0: String?) {}
            override fun onSetFailure(p0: String?) {}
        }, sdp)
    }

    fun handleRemoteIceCandidate(json: JSONObject) {
        val candidate = IceCandidate(
            json.getString("sdpMid"),
            json.getInt("sdpMLineIndex"),
            json.getString("candidate")
        )
        peerConnection?.addIceCandidate(candidate)
    }

    fun destroy() {
        localAudioTrack?.dispose()
        peerConnection?.close()
        peerConnectionFactory?.dispose()
    }
}

File: ios_app/WebRTCHandler.swift
import Foundation
import WebRTC

class WebRTCHandler: NSObject {
    private var factory: RTCPeerConnectionFactory
    private var peerConnection: RTCPeerConnection?
    private var localAudioTrack: RTCAudioTrack?

    var onOfferReady: ((String) -> Void)?
    var onIceCandidateReady: (([String: Any]) -> Void)?

    override init() {
        RTCInitializeSSL()
        let videoEncoderFactory = RTCDefaultVideoEncoderFactory()
        let videoDecoderFactory = RTCDefaultVideoDecoderFactory()
        self.factory = RTCPeerConnectionFactory(encoderFactory: videoEncoderFactory, decoderFactory: videoDecoderFactory)
        super.init()
    }

    func startCall() {
        print("Initiating WebRTC Audio Pipeline (Creating Offer)...")
        if peerConnection != nil {
            stopCall()
        }

        let config = RTCConfiguration()
        config.iceServers = [RTCIceServer(urlStrings: ["stun:stun.l.google.com:19302"])]

        let constraints = RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil)
        self.peerConnection = factory.peerConnection(with: config, constraints: constraints, delegate: self)

        createLocalAudioTrack()

        if let audioTrack = localAudioTrack {
            peerConnection?.add(audioTrack, streamIds: ["ARDAMS"])
        }

        let sdpConstraints = RTCMediaConstraints(mandatoryConstraints: [
            "OfferToReceiveAudio": "true",
            "OfferToReceiveVideo": "false"
        ], optionalConstraints: nil)

        peerConnection?.offer(for: sdpConstraints) { (sdp, error) in
            guard let sdp = sdp else { return }
            self.peerConnection?.setLocalDescription(sdp, completionHandler: { (error) in
                if error == nil {
                    print("SDP Request Generated. Ready to send.")
                    self.onOfferReady?(sdp.sdpDescription)
                }
            })
        }
    }

    private func createLocalAudioTrack() {
        let audioSource = factory.audioSource(with: nil)
        localAudioTrack = factory.audioTrack(with: audioSource, trackId: "ARDAMSa0")
        localAudioTrack?.isEnabled = true
        print("Local Audio Microphone Track Captured.")
    }

    func handleRemoteAnswer(sdpDescription: String) {
        let sdp = RTCSessionDescription(type: .answer, sdp: sdpDescription)
        peerConnection?.setRemoteDescription(sdp, completionHandler: { error in
            if error == nil {
                print("Remote Rescuer Answer received! Audio Bridged Successfully.")
            }
        })
    }

    func handleRemoteIceCandidate(json: [String: Any]) {
        guard let candidateStr = json["candidate"] as? String else { return }

        var sdpIndex: Int32 = 0
        if let idx = json["sdpMLineIndex"] as? Int32 {
            sdpIndex = idx
        } else if let idx = json["sdpMLineIndex"] as? Int {
            sdpIndex = Int32(idx)
        }

        let sdpMid = json["sdpMid"] as? String
        let candidate = RTCIceCandidate(sdp: candidateStr, sdpMLineIndex: sdpIndex, sdpMid: sdpMid)
        peerConnection?.add(candidate)
    }

    func stopCall() {
        print("WebRTC Terminating Command Received. Safely shutting down tunnel...")
        peerConnection?.close()
        peerConnection = nil
        localAudioTrack = nil
    }
}

extension WebRTCHandler: RTCPeerConnectionDelegate {
    func peerConnection(_ peerConnection: RTCPeerConnection, didGenerate candidate: RTCIceCandidate) {
        let json: [String: Any] = [
            "sdpMLineIndex": candidate.sdpMLineIndex,
            "sdpMid": candidate.sdpMid ?? "",
            "candidate": candidate.sdp
        ]
        onIceCandidateReady?(json)
    }

    func peerConnection(_ peerConnection: RTCPeerConnection, didChange stateChanged: RTCSignalingState) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove stream: RTCMediaStream) {}
    func peerConnectionShouldRenegotiate(_ peerConnection: RTCPeerConnection) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceConnectionState) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceGatheringState) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove candidates: [RTCIceCandidate]) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didOpen dataChannel: RTCDataChannel) {}
}
